pkg/operator/configobservation/apienablement/observe_runtime_config.go

```diff
@@ -12,11 +14,51 @@ import (
 	"github.com/openshift/library-go/pkg/operator/configobserver"
 	"github.com/openshift/library-go/pkg/operator/configobserver/featuregates"
 	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
 )
 
-var DefaultGroupVersionsByFeatureGate = map[configv1.FeatureGateName][]schema.GroupVersion{
-	"ValidatingAdmissionPolicy": {{Group: "admissionregistration.k8s.io", Version: "v1beta1"}},
-	"DynamicResourceAllocation": {{Group: "resource.k8s.io", Version: "v1alpha2"}},
+var defaultGroupVersionsByFeatureGate = map[configv1.FeatureGateName][]groupVersionByOpenshiftVersion{
+	"ValidatingAdmissionPolicy": {{GroupVersion: schema.GroupVersion{Group: "admissionregistration.k8s.io", Version: "v1beta1"}}},
+	"DynamicResourceAllocation": {
+		{KubeVersionRange: semver.MustParseRange("< 1.31.0"), GroupVersion: schema.GroupVersion{Group: "resource.k8s.io", Version: "v1alpha2"}},
+		{KubeVersionRange: semver.MustParseRange(">= 1.31.0"), GroupVersion: schema.GroupVersion{Group: "resource.k8s.io", Version: "v1alpha3"}},
+	},
+}
+
+type groupVersionByOpenshiftVersion struct {
+	schema.GroupVersion
+	KubeVersionRange semver.Range
+}
+
+func getGroupVersionByFeatureGate(groupVersionsByFeatureGate map[configv1.FeatureGateName][]groupVersionByOpenshiftVersion, kubeVersion semver.Version) (map[configv1.FeatureGateName][]schema.GroupVersion, error) {
+	result := make(map[configv1.FeatureGateName][]schema.GroupVersion, len(groupVersionsByFeatureGate))
+	groupByVersions := map[string][]string{}
+	for featureGate, APIGroups := range groupVersionsByFeatureGate {
+		for _, group := range APIGroups {
+			if group.KubeVersionRange == nil || group.KubeVersionRange(kubeVersion) {
+				groupByVersions[group.Group] = append(groupByVersions[group.Group], group.Version)
+				result[featureGate] = append(result[featureGate], group.GroupVersion)
+			}
+		}
+	}
+	var errs []error
+	for group, versions := range groupByVersions {
+		if len(versions) > 1 {
+			errs = append(errs, fmt.Errorf("found a duplicate group %v for a FeatureGates, versions found: %v", group, strings.Join(versions, ",")))
+		}
+	}
+	if len(errs) > 0 {
+		return nil, v1helpers.NewMultiLineAggregate(errs)
+	}
+	return result, nil
+}
+
+func GetDefaultGroupVersionByFeatureGate(kubeVersion semver.Version) map[configv1.FeatureGateName][]schema.GroupVersion {
+	result, err := getGroupVersionByFeatureGate(defaultGroupVersionsByFeatureGate, kubeVersion)
+	if err != nil {
+		panic(err)
```

Comment:
I think a panic would have been OK if we could detect intersecting version ranges at package initialization time, but since it turns out there's no practical way to do that, it's better to handle the error.

Reply:
@benluddy ack, passing the errors now